scripts/sync_protobuf/google_protobuf.py

```diff
@@ -0,0 +1,91 @@
+"""
+Generates the protobuf stubs for the given protobuf version using mypy-protobuf.
+Generally, new minor versions are a good time to update the stubs.
+"""
+
+from __future__ import annotations
+
+import json
+import re
+import shutil
+import subprocess
+import sys
+import tempfile
+from pathlib import Path
+
+from _utils import MYPY_PROTOBUF_VERSION, REPO_ROOT, download_file, extract_archive, run_protoc, update_metadata
+
+# Whenever you update PACKAGE_VERSION here, version should be updated
+# in stubs/protobuf/METADATA.toml and vice-versa.
+PACKAGE_VERSION = "27.1"
+
+STUBS_FOLDER = REPO_ROOT / "stubs" / "protobuf"
+ARCHIVE_FILENAME = f"protobuf-{PACKAGE_VERSION}.zip"
+ARCHIVE_URL = f"https://github.com/protocolbuffers/protobuf/releases/download/v{PACKAGE_VERSION}/{ARCHIVE_FILENAME}"
+EXTRACTED_PACKAGE_DIR = f"protobuf-{PACKAGE_VERSION}"
+
+VERSION_PATTERN = re.compile(r'def game_version\(\):\n    return "(.+?)"')
+PROTO_FILE_PATTERN = re.compile(r'"//:(.*)_proto"')
+
+
+def extract_python_version(file_path: Path) -> str:
+    """Extract the Python version from https://github.com/protocolbuffers/protobuf/blob/main/version.json"""
+    with open(file_path) as file:
+        data: dict[str, dict[str, dict[str, str]]] = json.load(file)
+    # The root key will be the protobuf source code version
+    return next(iter(data.values()))["languages"]["python"]
```

Comment:
I'd like to see some validation of the version, considering its coming from an outside source. Something like:

```suggestion
        data = json.load(file)
        # The root key will be the protobuf source code version
        version = next(iter(data.values()))["languages"]["python"]
        assert isinstance(version, str)
        assert re.fullmatch(r"...", version)  # proper re here
        return version
```

This way we're also sure (at runtime) that version has the correct type and format.

Reply:
I feel like validating the version string is unnecessary extra work. We're not doing anything with it other than displaying it. Proper validation should probably use a Python packaging library (I don't remember which)
The str assertion I still find valuable in case protobuff changes the structure of that file and the value becomes an object (dict)